File: public/planet.js
'use strict'

var so = so || {};

var halfPI = Math.PI/2;
var quarterPI = Math.PI/4;
var tau = Math.PI * 2;

so.Planet = function( _camera, _radius, _position, _segments, _fov, _screenWidth, renderer ) {

	var me = this;
	
	var fragmentShader;
	var vertexShader;

	renderer = renderer ? renderer : new THREE.WebGLRenderer();

	var position = new THREE.Vector3(0,0,0);

	me.obj = new THREE.Object3D();
	me.obj.position = _position || new THREE.Vector3();

	//trying to solve clipping issue.
//	var placeHolder = new THREE.Mesh( new THREE.SphereGeometry( radius * 1.3 ) );
//	placeHolder.position.z -= radius;
//	me.obj.add( placeHolder );

	var camera = _camera;
	var radius = _radius || 6353000;
	var segments = _segments || 20;
	
	var fov = _fov || 30;
	fov = fov * .0174532925;//Convert to radians

	var textureProvider = new so.TextureProvider( renderer, radius, 42 );

	var screenWidth = _screenWidth || 768;
	//tan of fov/screenWidth is first half of pixel size on planet calc
	var vs = Math.tan(fov/screenWidth);
	
	function findClipMapCount(){
		var minTheta = getMinTheta(radius, 2);//smallest theta is 2 units of height
		var i = 0;
		var theta = 100;
		while( theta > minTheta){
			i++;
			theta = (1 / Math.pow(2,i) ) * Math.PI;
		}
		var min = 21;
		return i < min ? i : min;
	}

	//Don't resond to update unless init has completed
	var inited = false;


	//load shaders before init
	var fileGetter = new so.FileLoader();
	fileGetter.getFiles(['fragmentShader.glsl', 'vertexShader.glsl', 'wireframeFragmentShader.glsl'], init);

	
	function init( files ) {

		fragmentShader = files['fragmentShader.glsl'];
//		fragmentShader = files['wireframeFragmentShader.glsl'];
		vertexShader = files['vertexShader.glsl'];
		inited = true;
		initClipMaps();

	}
	
	var clipMapCount = findClipMapCount();
	
	clipMapCount = 20;	
	
	var circleGeo = new THREE.RingGeometry( .000001, radius,  segments, segments, 0, tau ); 


/*
 *
 * Update loop
 *
 */

	var clock = new THREE.Clock(), localCam, cameraDistance, delta = 0, theta, phi, maxTheta, minTheta;
	var lockHeight = 2;
	
	me.update = function( ) {
		logText = '';

		if(! inited){ return; }
		
		delta += clock.getDelta();
	
		if(delta >= 1){

			var tMesh = me.obj.clone();
			tMesh.position = tMesh.localToWorld(tMesh.position);
			tMesh.position.z -= radius;
			tMesh.updateMatrixWorld(false);

			localCam = camera.position.clone();
			tMesh.worldToLocal(localCam);
			cameraDistance = camera.position.distanceTo(tMesh.position) - radius;


			getTheta(localCam.x, localCam.y, localCam.z );
			getPhi(localCam.x, localCam.z );
			minTheta =  getMinTheta( radius, cameraDistance );
			maxTheta = getMaxTheta( radius, cameraDistance );

			getLockHeight( cameraDistance );	

			//Determine if we need to update clipmaps and to which phi, theta

			var m = new THREE.Matrix4();
			var p = new THREE.Vector3();
			
			m.lookAt(localCam, p, new THREE.Vector3(0,1,0) );
			var tr = m.decompose()[ 1 ].inverse();

			updateClipMaps(lockHeight, tr, phi, theta);

			log('lockHeight', lockHeight);
			log("planetpos",me.obj.position);
			log("adjusted planetpos",tMesh.position);
			log('radius', radius);
			log('height', cameraDistance);
			log("actualcam",camera.position);
			log("localcam",localCam);
			log("theta", theta);
			log("phi", phi);
			log('minTheta', minTheta);
			log('maxTheta',maxTheta);
			log('clipMapCount', clipMapCount+1);
			$('#info').html(logText);
			delta = 0;
		}
	}
	
	function getLockHeight( height ) {
		var max = 30, min = 1, midpoint = Math.floor( max / 2), step = midpoint;

		while( 1 ) {
			step = Math.round( step / 2 );
			step = step == 0 ? 1 : step;
			lockHeight = Math.pow(2,midpoint);

			if( height < 2 ) {//If we have a negative height, whups
				lockHeight = 2;
				break;
			}

			if ( height >= lockHeight ) {
				if( midpoint >= max || height < Math.pow(2, midpoint+1) ) {
					break;
				} else {
					midpoint += step ;
				}
			} else {
					midpoint -= step ;
			}
		}
	}
	
/*
 *
 * Update Clipmaps
 *
 */
	var clipMaps = [], i;
	var colors = [0xFF0000, 0x0000FF, 0x00FF00];//red, blue, green
	function updateClipMaps( height, rotate ) {
		
		//min theta planet pixel size / radius i minimum theta
		//max theta
		for( var i = 0; i < clipMapCount; i++ ) {
			if( clipMaps[i].visible === false ) {
				if( clipMaps[i].theta < maxTheta && clipMaps[i].theta > minTheta ) {
					me.obj.add(clipMaps[i].mesh);
					clipMaps[i].visible = true;
				}
			} else {
				if( clipMaps[i].theta < minTheta || clipMaps[i].theta > maxTheta ) {
					me.obj.remove(clipMaps[i].mesh);
					clipMaps[i].visible = false;
					continue;
				}
			}
			if(clipMaps[i].visible) {
				log('level: ' + i , ' theta:' + clipMaps[i].theta);
				clipMaps[i].material.uniforms.meshRotation.value = rotate ;
				//clipMaps[i].material.uniforms.texture = textureProvider.getTexture( i, phi, theta ); 
				if(i+1 === clipMapCount || clipMaps[i+1].theta < minTheta ){
					clipMaps[i].material.uniforms.last.value =  1;
				}else{
					clipMaps[i].material.uniforms.last.value =  0;
				}
			}
		}
	}

	
	function initClipMaps( ) {

		clipMaps.length = 0;//empty array of any other clipMaps in case we've been re-init'd runtime
	
		var t = quarterPI;
		var scale, scaledPI;
		for( i = 0; i < clipMapCount; i++ ) {

			scale = ( 1 / Math.pow( 2, i+1 ) ) ;
			scaledPI = Math.PI /  2 * scale ;
			clipMaps[i] = {};

			clipMaps[i].material = new THREE.ShaderMaterial( {
				uniforms: { 
					texture: { // texture in slot 0, loaded with ImageUtils
						type: "t", 
						value: undefined
					},  
					meshRotation: { 
						type: "v4",
						value: new THREE.Vector4(0,0,0,0),
					},
					icolor: {
						type: "c",
						value: new THREE.Color(colors[i % 3]),
					},
					scaledPI: {
						type: "f",
						value: scaledPI 
					},
					radius: {
						type: "f",
						value: radius 
					},
					last: {
						type: "i",
						value: 0 
					},
				},

				vertexShader: vertexShader,
				fragmentShader: fragmentShader

			} );	

			clipMaps[i].theta = t;
			clipMaps[i].mesh = new THREE.Mesh(circleGeo, clipMaps[i].material);
			clipMaps[i].visible = false;

			t /= 2;//Each successive clipMap covers half as much theta
		}
	}



/*
 *
 * Helper functions
 * 
 */
	function getMaxTheta( radius, height ) {
		var mt = Math.acos( radius / (radius + height ) );	
		return mt < halfPI ? mt : halfPI;
	}
	
	function getMinTheta( radius, height ) {
		var lt = ( (height * vs) / radius ) * segments;//multiply by segments because this is theta per triangle
		return lt < quarterPI ? lt : quarterPI;
	}

	function getTheta( x, y, z) {
		//q =tan-1(y/(z2+x2)1/2)
		theta = Math.PI / 2 - Math.atan( y / Math.pow( z * z + x * x, .5 ) );
	}

	function getPhi( x,z ) {
		//f = tan-1(x/z).
		phi = Math.atan( localCam.x / localCam.z );
		//Now adjust for special cases
		if(localCam.x < 0 && localCam.z < 0){
			phi -= Math.PI;
		}else if(localCam.z < 0){
			phi += Math.PI;
		}
	}

	//Simple log function
	var logText;
	function log(s, t){
		if(t instanceof THREE.Vector3){
			logText += s + ": x " + t.x + ", y " + t.y + ", z " + t.z; 
		}else{
			logText += s + ": " + t;
		}
		logText += "<br />";
	}
};

File: Assets/Scripts/Planet.cs
using UnityEngine;
using System.Collections;

public class Planet : MonoBehaviour
{

	static float halfPI = Mathf.PI / 2f;
	static float quarterPI = Mathf.PI / 4f;
	
//	public Camera camera;
	public int radius = 6353000;
	public int segments = 64;
	public int screenWidth = Screen.width;
	public int minHeight = 2;
	public float fov = 30f;
	public float smallestTheta;
	public ClipMap clipMapPrefab;
	public float ClipMapUpdateSpeed = .1f;
	
	private Camera playerCamera;
	private bool inited = false;
	private float vs; //minimum view angle
	private int clipMapCount;
	private ClipMap[] clipMaps;
	private float[] scaledPI;
	private float delta = 0;
	private float oldHeightLock = 0, oldPhiLock = 0, oldThetaLock = 0;
	
	// Use this for initialization
	public void Init (Camera _camera, int _radius, Vector3 _position, int _segments, int _fov)
	{
		playerCamera = GameObject.FindWithTag ("MainCamera").GetComponent<Camera>();
		
		fov = _fov == 0 ? _fov : fov;
		fov = (fov * Mathf.PI) / 180;// .0174532925f;//Convert to radians
		playerCamera = _camera != null ? _camera : playerCamera;	
		radius = _radius == 0 ? _radius : radius;
		Transform t = gameObject.GetComponent<Transform> ();	
		t.position = _position;
		segments = _segments == 0 ? _segments : segments;
		vs = Mathf.Tan (fov / screenWidth);	
		
		smallestTheta = smallestTheta != 0 ? smallestTheta : getMinTheta (radius, minHeight);
		setClipMapCount ();	
		
		initClipMaps ();
		
		inited = true;	
	}
	
	
	// Update is called once per frame
	public void Update ()
	{
		if (!inited)
			return;
		delta += Time.deltaTime;
		
		if (delta > ClipMapUpdateSpeed) {
			Transform t = gameObject.transform;
			t.position.Set (t.position.x, t.position.y, t.position.z - radius);
			Vector3 localCameraPosition = t.InverseTransformPoint (playerCamera.transform.position);	
			float cameraDistance = Vector3.Distance (t.position, playerCamera.transform.position) - radius;	
		
			float theta = getTheta (localCameraPosition);
			float phi = getPhi (localCameraPosition);
			
			float heightLock = getHeightLock (cameraDistance);
			float minTheta = getMinTheta (radius, heightLock);
			
			float phiLock = getPhiLock (minTheta, phi);

			float thetaLock = getThetaLock (minTheta, theta);

			//If the a locked location has changed, update the clipmaps
			if (oldHeightLock != heightLock || oldPhiLock != phiLock || oldThetaLock != thetaLock) {
				float maxTheta = getMaxTheta (radius, heightLock);
				
				Quaternion pq = Quaternion.AngleAxis (Utils.RadiansToDegrees (-phiLock), new Vector3 (0, 1, 0));
				Quaternion tq = Quaternion.AngleAxis (Utils.RadiansToDegrees ((Mathf.PI / 2) - thetaLock), new Vector3 (1, 0, 0));
		
				updateClipMaps (heightLock, pq * tq, minTheta, maxTheta);	
				
				oldHeightLock = heightLock;
				oldPhiLock = phiLock;
				oldThetaLock = thetaLock;
			}	
				
			t.position.Set (t.position.x, t.position.y, t.position.z + radius);
			delta = 0f;
		}
		
	}

	private void updateClipMaps (float heightLock, Quaternion rotation, float minTheta, float maxTheta)
	{
		
		for (var i = 0; i< clipMapCount; i++) {
			if (clipMaps [i].visible == false) {
				if (clipMaps [i].theta < maxTheta && clipMaps [i].theta > minTheta) {
					clipMaps [i].visible = true;
				}
			} else {
				if (clipMaps [i].theta < minTheta || clipMaps [i].theta > maxTheta) {
					clipMaps [i].visible = false;
					continue;
				}
			}
			if (clipMaps [i].visible) {
				clipMaps [i].rotate = rotation;
				if(i+1 == clipMapCount || clipMaps[i+1].theta < minTheta ){
		          clipMaps[i].last =  true;
		        }else{
    		      clipMaps[i].last =  false;
		        }
			}
		}
		
	}
	
	private void initClipMaps ()
	{
		
		clipMaps = new ClipMap[clipMapCount];	
		float t = quarterPI;
		float scale;
		scaledPI = new float[clipMapCount];
		
		for (int i = 0; i < clipMapCount; i++) {
			scale = (1 / Mathf.Pow (2, i + 1));
			scaledPI [i] = Mathf.PI / 2 * scale;
			clipMaps [i] = (ClipMap)Instantiate (clipMapPrefab);
			clipMaps [i].Init (scaledPI [i], radius);
			clipMaps [i].theta = t;
			t /= 2f;
		}
		
	}
	
	private void setClipMapCount ()
	{
		float theta = 100;
		while (theta > smallestTheta) {
			theta = (1 / Mathf.Pow (2, ++clipMapCount)) * Mathf.PI;
		}
	}

	private float getHeightLock (float height)
	{
		float max = 30, midpoint = Mathf.Round (max / 2f), step = midpoint;
		float heightLock = 2;
		
		while (true) {
			step = Mathf.Round (step / 2);
			step = step == 0 ? 1 : step;
			heightLock = Mathf.Pow (2, midpoint);

			if (height < 2) {//If we have a negative height, whups
				heightLock = 2;
				break;
			}

			if (height >= heightLock) {
				if (midpoint >= max || height < Mathf.Pow (2, midpoint + 1)) {
					break;
				} else {
					midpoint += step;
				}
			} else {
				midpoint -= step;
			}
		}	
		return heightLock;	
	}

	private float getMinTheta (float _radius, float height)
	{ 
		float lt = ((height * vs) / _radius) * segments;//multiply by segments because this is theta per triangle
		lt = lt < quarterPI ? lt : quarterPI;
		return lt < 0 ? smallestTheta : lt; 
	}

	private float getMaxTheta (float radius, float height)
	{
		float mt = Mathf.Acos (radius / (radius + height));
		return mt < halfPI ? mt : halfPI;	
	}
	
	private float getThetaLock (float minTheta, float theta)
	{
		float max = Mathf.PI / minTheta, midpoint = max / 2, step = midpoint, thetaLock;
		while (true) {
			step = step / 2f;
			step = step == 0f ? 1f : step;

			thetaLock = minTheta * midpoint;

			if (theta >= thetaLock) {
				if (theta < minTheta * (midpoint + 1)) {
					break;
				} else {
					midpoint += step;
				}
			} else {
				midpoint -= step;
			}
		}
		return thetaLock;
	}
	
	private float getTheta (Vector3 v)
	{
		//q =tan-1(y/(z2+x2)1/2)
		return Mathf.PI / 2f - Mathf.Atan (v.y / Mathf.Pow ((v.z * v.z + v.x * v.x), .5f));
	}

	private float getPhi (Vector3 v)
	{
		//f = tan-1(x/z).
		float phi = Mathf.Atan (v.x / v.z);
		//Now adjust for special cases
		if (v.x < 0 && v.z < 0) {
			phi -= Mathf.PI;
		} else if (v.z < 0) {
			phi += Mathf.PI;
		}
		return phi;	
	}

	private float getPhiLock (float minTheta, float phi)
	{
		float max = Mathf.PI * 2 / minTheta, midpoint = max / 2, step = midpoint, tphi = phi + Mathf.PI, phiLock;
		while (true) {
			step = step / 2;
			step = step == 0 ? 1 : step;

			phiLock = minTheta * midpoint;

			if (tphi >= phiLock) {
				if (tphi < minTheta * (midpoint + 1)) {
					break;
				} else {
					midpoint += step;
				}
			} else {
				midpoint -= step;
			}
		}
		phiLock -= Mathf.PI;	
		return phiLock;
	}
}
